src/ui/src/js/services/garden_service.js

```diff
@@ -89,22 +116,184 @@ export default function gardenService($rootScope, $http) {
+
+    return allHttpFieldsEmpty;
+  };
+
+  GardenService.formToServerModel = function(data, model) {
+    /* Carefully pick apart the form data and translate it to the correct server
+     * model. Throw an error if the entire form is empty (i.e., don't allow
+     * empty connection parameters for both entry points on a remote garden).
+     */
+    const {connection_type: modelConnectionType, ...modelWithoutConxType} = model;
+    let newModel = {...data};
+    newModel['connection_type'] = modelConnectionType;
+
+    const updatedConnectionParams = {};
+    const emptyConnections = {};
+    const emptyChecker = {
+      'stomp': isEmptyStompConnection,
+      'http': isEmptyHttpConnection,
+    };
+
+    for (const modelEntryPointName of Object.keys(modelWithoutConxType)) {
+      // modelEntryPointName is either 'http' or 'stomp'
+      const modelEntryPointMap = modelWithoutConxType[modelEntryPointName];
+      const isEmpty = emptyChecker[modelEntryPointName](modelEntryPointMap);
+
+      if (isEmpty) {
+        emptyConnections[modelEntryPointName] = true;
+        continue;
+      } else {
+        emptyConnections[modelEntryPointName] = false;
+      }
+
+      const updatedEntryPoint = {};
+
+      for (const modelEntryPointKey of Object.keys(modelEntryPointMap)) {
+        const modelEntryPointValue = modelEntryPointMap[modelEntryPointKey];
+
+        if (modelEntryPointName === 'stomp' && modelEntryPointKey === 'headers') {
+          // the ugly corner case is the stomp headers
+          const formStompHeaders = modelEntryPointValue;
+          const modelUpdatedStompHeaderArray = [];
+
+          for (const formStompHeader of formStompHeaders) {
+            const formStompKey = formStompHeader['key'];
+            const formStompValue = formStompHeader['value'];
+
+            // assume that we have both a key and a value or neither
+            if (formStompKey && formStompValue) {
+              modelUpdatedStompHeaderArray.push(
+                  {
+                    'key': formStompKey,
+                    'value': formStompValue,
+                  },
+              );
+            }
+          }
+
+          if (modelUpdatedStompHeaderArray.length > 0) {
+            updatedEntryPoint['headers'] = modelUpdatedStompHeaderArray;
+          }
+        } else {
+          updatedEntryPoint[modelEntryPointKey] = modelEntryPointValue;
         }
-        model['connection_params'][field] = stompHeaders;
-      } else if (field != 'connection_type') {
-        model['connection_params'][field] = form[field];
       }
+      updatedConnectionParams[modelEntryPointName] = updatedEntryPoint;
     }
 
-    return model;
+    if (emptyConnections['http'] && emptyConnections['stomp']) {
+      throw Error('One of \'http\' or \'stomp\' connection must be defined');
+    } else if (emptyConnections['http'] && modelConnectionType === 'HTTP') {
+      throw Error('Connection type is \'HTTP\' but http connection parameters' +
```

Comment:
I lobbied for doing away with this validation in another comment, but in the event that this message sticks around, there's a space missing between "parameters" and "are".

Reply:
This is removed.